src/bytearray.rs

```diff
@@ -0,0 +1,236 @@
+    fn borrow(&self) -> &Bytes {
+        Bytes::new(&self.bytes)
+    }
+}
+
+impl<const N: usize> BorrowMut<Bytes> for ByteArray<N> {
+    fn borrow_mut(&mut self) -> &mut Bytes {
+        unsafe { &mut *(&mut self.bytes as &mut [u8] as *mut [u8] as *mut Bytes) }
+    }
+}
+
+impl<Rhs, const N: usize> PartialEq<Rhs> for ByteArray<N>
+where
+    Rhs: ?Sized + Borrow<[u8; N]>,
+{
+    fn eq(&self, other: &Rhs) -> bool {
+        self.as_ref().eq(other.borrow())
+    }
+}
+
+impl<Rhs, const N: usize> PartialOrd<Rhs> for ByteArray<N>
+where
+    Rhs: ?Sized + Borrow<[u8; N]>,
+{
+    fn partial_cmp(&self, other: &Rhs) -> Option<Ordering> {
+        self.as_ref().partial_cmp(other.borrow())
+    }
+}
+
+impl<const N: usize> Hash for ByteArray<N> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.bytes.hash(state);
+    }
+}
+
+impl<const N: usize> IntoIterator for ByteArray<N> {
+    type Item = u8;
+    type IntoIter = <[u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIterator::into_iter(self.bytes)
+    }
+}
+
+impl<'a, const N: usize> IntoIterator for &'a ByteArray<N> {
+    type Item = &'a u8;
+    type IntoIter = <&'a [u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.bytes.iter()
+    }
+}
+
+impl<'a, const N: usize> IntoIterator for &'a mut ByteArray<N> {
+    type Item = &'a mut u8;
+    type IntoIter = <&'a mut [u8; N] as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.bytes.iter_mut()
+    }
+}
+
+impl<const N: usize> Serialize for ByteArray<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.bytes)
+    }
+}
+
+struct ByteArrayVisitor<const N: usize>;
+
+impl<'de, const N: usize> Visitor<'de> for ByteArrayVisitor<N> {
+    type Value = ByteArray<N>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "[u8; {}]", N)
+    }
+
+    fn visit_seq<V>(self, mut visitor: V) -> Result<ByteArray<N>, V::Error>
+    where
+        V: SeqAccess<'de>,
+    {
+        let mut bytes = [0; N];
+        let mut idx = 0;
+        while let Some(b) = visitor.next_element()? {
+            bytes[idx] = b;
+            idx += 1;
+            if idx == N {
+                break;
+            }
+        }
+
+        if idx != N {
+            return Err(V::Error::invalid_length(N, &self));
+        }
+
+        Ok(ByteArray::from(bytes))
+    }
+
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<ByteArray<N>, E>
+    where
+        E: Error,
+    {
+        Ok(ByteArray {
+            bytes: v
+                .try_into()
+                .map_err(|_| E::invalid_length(v.len(), &self))?,
+        })
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<ByteArray<N>, E>
+    where
+        E: Error,
+    {
+        Ok(ByteArray {
```

Comment:
Nit: For simplicity: consider `self.visit_bytes(v.as_bytes())`

Reply:
Done in [4c24ac8](https://github.com/serde-rs/bytes/pull/28/commits/4c24ac8dae6f92319ccbabc26c4c8959c57b57f7)